File: vane-core/src/main/java/org/oddlama/vane/util/MaterialUtil.java
package org.oddlama.vane.util;

import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.Registry;

public class MaterialUtil {

    public static Material materialFrom(NamespacedKey key) {
        return Registry.MATERIAL.get(key);
    }

    public static boolean isSeededPlant(Material type) {
        switch (type) {
            default:
                return false;
            case WHEAT:
            case CARROTS:
            case POTATOES:
            case BEETROOTS:
            case NETHER_WART:
                return true;
        }
    }

    public static Material seedFor(Material plantType) {
        switch (plantType) {
            default:
                return null;
            case WHEAT:
                return Material.WHEAT_SEEDS;
            case CARROTS:
                return Material.CARROT;
            case POTATOES:
                return Material.POTATO;
            case BEETROOTS:
                return Material.BEETROOT_SEEDS;
            case NETHER_WART:
                return Material.NETHER_WART;
        }
    }

    public static Material farmlandFor(Material seedType) {
        switch (seedType) {
            default:
                return null;
            case WHEAT_SEEDS:
            case CARROT:
            case POTATO:
            case BEETROOT_SEEDS:
                return Material.FARMLAND;
            case NETHER_WART:
                return Material.SOUL_SAND;
        }
    }

    public static boolean isReplaceableGrass(Material type) {
        switch (type) {
            default:
                return false;
            case TALL_GRASS:
            case SHORT_GRASS:
                return true;
        }
    }

    public static boolean isTillable(Material type) {
        switch (type) {
            default:
                return false;
            case DIRT:
            case GRASS_BLOCK:
            case DIRT_PATH:
                return true;
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/util/MaterialUtil.kt
package org.oddlama.vane.util

import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.Registry

/**
 * Utility helpers for resolving and classifying materials.
 */
object MaterialUtil {
    /**
     * Resolves a material from its namespaced key.
     */
    @JvmStatic
    fun materialFrom(key: NamespacedKey): Material? = Registry.MATERIAL.get(key)

    /**
     * Returns whether a block material is a seeded crop plant.
     */
    @JvmStatic
    fun isSeededPlant(type: Material): Boolean = when (type) {
        Material.WHEAT, Material.CARROTS, Material.POTATOES, Material.BEETROOTS, Material.NETHER_WART -> true
        else -> false
    }

    /**
     * Returns the seed item for a planted crop block material.
     */
    @JvmStatic
    fun seedFor(plantType: Material): Material? = when (plantType) {
        Material.WHEAT       -> Material.WHEAT_SEEDS
        Material.CARROTS     -> Material.CARROT
        Material.POTATOES    -> Material.POTATO
        Material.BEETROOTS   -> Material.BEETROOT_SEEDS
        Material.NETHER_WART -> Material.NETHER_WART
        else                 -> null
    }

    /**
     * Returns the farmland-like block required by a seed item.
     */
    @JvmStatic
    fun farmlandFor(seedType: Material): Material? = when (seedType) {
        Material.WHEAT_SEEDS, Material.CARROT, Material.POTATO, Material.BEETROOT_SEEDS -> Material.FARMLAND
        Material.NETHER_WART -> Material.SOUL_SAND
        else                 -> null
    }

    /**
     * Returns whether the material is replaceable grass.
     */
    fun isReplaceableGrass(type: Material): Boolean = when (type) {
        Material.TALL_GRASS, Material.SHORT_GRASS -> true
        else -> false
    }

    /**
     * Returns whether the material can be tilled by a hoe.
     */
    @JvmStatic
    fun isTillable(type: Material): Boolean = when (type) {
        Material.DIRT, Material.GRASS_BLOCK, Material.DIRT_PATH -> true
        else -> false
    }
}
